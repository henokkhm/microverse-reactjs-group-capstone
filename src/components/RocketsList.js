import { useSelector } from 'react-redux';

import Rocket from './Rocket';
import styles from '../styles/RocketList.module.css';

function RocketsList() {
  const { rocketsList, isLoadingRockets, errorLoadingRockets } = useSelector(
    (store) => store.rockets,
  );

  if (isLoadingRockets) {
    return <div className={styles.loading}>Loading rockets...</div>;
  }

  if (errorLoadingRockets) {
    return (
      <div className={styles.error}>
        Something went wrong! Please try again later.
      </div>
    );
  }

  return (
    <ul className={styles.rocketList}>
      {rocketsList.map((rocket) => (
        <Rocket
          key={rocket.id}
          id={rocket.id}
          name={rocket.name}
          description={rocket.description}
          imgUrl={rocket.flickr_images[0]}
          isReserved={rocket.isReserved}
        />
      ))}
    </ul>
  );
}

export default RocketsList;
