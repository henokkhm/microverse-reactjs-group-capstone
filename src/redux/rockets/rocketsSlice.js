import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import getRockets from '../../api/rockets-api';

const initialState = {
  rocketsList: [],
};

export const getRocketsFromAPI = createAsyncThunk(
  'rockets/getRocketsFromAPI',
  async (_, thunkAPI) => {
    try {
      const resp = await getRockets();
      const rocketsData = resp.map((rocket) => ({
        id: rocket.id,
        name: rocket.name,
        type: rocket.type,
        flickr_images: rocket.flickr_images,
      }));
      return rocketsData;
    } catch (error) {
      return thunkAPI.rejectWithValue(`Something went wrong! ${error}`);
    }
  },
);

const rocketsSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(getRocketsFromAPI.fulfilled, (state, action) => {
      state.rocketsList = action.payload;
    });
  },
});

export default rocketsSlice.reducer;
