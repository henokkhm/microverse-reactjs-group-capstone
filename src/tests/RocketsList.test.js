import fetchMock from 'jest-fetch-mock';
import '@testing-library/jest-dom';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

import RocketsList from '../components/RocketsList';
import { getRocketsFromAPI } from '../redux/rockets/rocketsSlice';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('RocketsList component', () => {
  afterEach(() => {
    fetchMock.resetMocks();
  });

  test('renders as expected', async () => {
    // ARRANGE
    const mockRockets = [
      {
        id: 'id-1',
        name: '[Mock-data] Falcon 1',
        description: 'Mock Description Falcon 1',
        flickr_images: ['https://imgur.com/DaCfMsj.jpg'],
        isReserved: false,
      },
      {
        id: 'id-2',
        name: '[Mock-data] Falcon 9',
        description: 'Mock Description Falcon 9',
        flickr_images: [
          'https://farm1.staticflickr.com/929/28787338307_3453a11a77_b.jpg',
        ],
        isReserved: false,
      },
    ];

    fetchMock.mockResponse(JSON.stringify(mockRockets));

    const store = mockStore({
      rockets: {
        rocketsList: mockRockets,
        isLoadingRockets: false,
        errorLoadingRockets: false,
      },
    });

    await store.dispatch(getRocketsFromAPI());

    // ACT
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <RocketsList />
        </BrowserRouter>
      </Provider>,
    );

    // ASSERT
    expect(asFragment()).toMatchSnapshot();
  });
});
