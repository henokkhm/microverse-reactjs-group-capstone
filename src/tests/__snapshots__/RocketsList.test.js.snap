// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RocketsList component renders as expected 1`] = `
<DocumentFragment>
  <ul
    class="rocketList"
  >
    <li
      class="rocketCard"
    >
      <div
        class="rocketImgWrapper"
      >
        <img
          alt="[Mock-data] Falcon 1 rocket"
          class="rocketImg"
          src="https://imgur.com/DaCfMsj.jpg"
        />
      </div>
      <div
        class="rocketCardBody"
      >
        <h3
          class="rocketName"
        >
          [Mock-data] Falcon 1
        </h3>
        <p
          class="rocketDescription"
        >
          Mock Description Falcon 1
        </p>
        <button
          class="reserveRocketBtn"
          type="button"
        >
          Reserve Rocket
        </button>
      </div>
    </li>
    <li
      class="rocketCard"
    >
      <div
        class="rocketImgWrapper"
      >
        <img
          alt="[Mock-data] Falcon 9 rocket"
          class="rocketImg"
          src="https://farm1.staticflickr.com/929/28787338307_3453a11a77_b.jpg"
        />
      </div>
      <div
        class="rocketCardBody"
      >
        <h3
          class="rocketName"
        >
          [Mock-data] Falcon 9
        </h3>
        <p
          class="rocketDescription"
        >
          Mock Description Falcon 9
        </p>
        <button
          class="reserveRocketBtn"
          type="button"
        >
          Reserve Rocket
        </button>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
